src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -197,16 +201,149 @@ class SubcategorySelectionPools<SubgraphPool> {
+
+  const v2Pools = (await v2SubgraphProvider.getPools(tokenIn, tokenOut, {
+    blockNumber,
+  })).sort((a, b) => b.reserve - a.reserve);
+  const v3Pools = (await v3SubgraphProvider.getPools(tokenIn, tokenOut, {
+    blockNumber,
+  })).sort((a, b) => b.tvlUSD - a.tvlUSD);
+
+  for (const pool of v2Pools) {
+    pool.token0.id = pool.token0.id.toLowerCase();
+    pool.token1.id = pool.token1.id.toLowerCase();
+  }
+  for (const pool of v3Pools) {
+    pool.token0.id = pool.token0.id.toLowerCase();
+    pool.token1.id = pool.token1.id.toLowerCase();
+  }
+
+  const tokenInAddress = tokenIn.address.toLowerCase();
+  const tokenOutAddress = tokenOut.address.toLowerCase();
+
+  const v2SelectedPools: {
+    forTokenIn?: V2SubgraphPool;
+    forTokenOut?: V2SubgraphPool;
+  } = {};
+  const previouslySelectedV2PoolIds = new Set(v2Candidates?.subgraphPools.map(pool => pool.id) ?? []);
+
+  const crossTokenAgainstV3TokenOut = v3Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token0.id.toLowerCase() === tokenOutAddress ?
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token1.id.toLowerCase() :
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token0.id.toLowerCase();
+  const crossTokenAgainstV3TokenIn = v3Candidates?.candidatePools.selections.topByTVLUsingTokenIn[0]?.token0.id.toLowerCase() === tokenInAddress ?
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenIn[0]?.token1.id.toLowerCase() :
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenIn[0]?.token0.id.toLowerCase();
+
+  for (const pool of v2Pools) {
+    if (previouslySelectedV2PoolIds.has(pool.id.toLowerCase())) {
+      continue;
+    }
+
+    if (v2SelectedPools.forTokenIn !== undefined && v2SelectedPools.forTokenOut !== undefined) {
+      break;
+    }
+
+    if (
+      v2SelectedPools.forTokenIn === undefined && (
+        (pool.token0.id.toLowerCase() === tokenOutAddress && pool.token1.id.toLowerCase() === crossTokenAgainstV3TokenIn) ||
+        (pool.token1.id.toLowerCase() === tokenOutAddress && pool.token0.id.toLowerCase() === crossTokenAgainstV3TokenIn)
+      )
+    ) {
+      v2SelectedPools.forTokenIn = pool;
+    }
+
+    if (
+      v2SelectedPools.forTokenOut === undefined && (
+        (pool.token0.id.toLowerCase() === tokenInAddress && pool.token1.id.toLowerCase() === crossTokenAgainstV3TokenOut) ||
+        (pool.token1.id.toLowerCase() === tokenInAddress && pool.token0.id.toLowerCase() === crossTokenAgainstV3TokenOut)
+      )
+    ) {
+      v2SelectedPools.forTokenOut = pool;
+    }
+  }
+
+  const v3SelectedPools: {
+    forTokenIn?: V3SubgraphPool;
+    forTokenOut?: V3SubgraphPool;
+  } = {};
+  const previouslySelectedV3PoolIds = new Set(v3Candidates?.subgraphPools.map(pool => pool.id) ?? []);
+
+  const crossTokenAgainstV2TokenOut = v2Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token0.id.toLowerCase() === tokenOutAddress ?
```

Comment:
this logic is the exact same with the one above with v2 and v3 switch, can we make them a function and consolidate?

Reply:
let me see if I can do that